crates/astria-proto/src/native/sequencer/v1alpha1.rs

```diff
@@ -3,18 +3,592 @@ use std::{
+            }
+            SignedTransactionErrorKind::VerificationKey(_) => {
+                "could not reconstruct an ed25519 verification key from the bytes contained in the \
+                 `public_key` field of the raw protobuf message"
+            }
+            SignedTransactionErrorKind::Verification(_) => {
+                "the encoded bytes of the raw unsigned protobuf transaction could not be verified"
+            }
+        };
+        f.pad(msg)
+    }
+}
+
+impl Error for SignedTransactionError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match &self.kind {
+            SignedTransactionErrorKind::UnsetTransaction => None,
+            SignedTransactionErrorKind::Signature(e)
+            | SignedTransactionErrorKind::VerificationKey(e)
+            | SignedTransactionErrorKind::Verification(e) => Some(e),
+            SignedTransactionErrorKind::Transaction(e) => Some(e),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum SignedTransactionErrorKind {
+    UnsetTransaction,
+    Signature(ed25519_consensus::Error),
+    Transaction(UnsignedTransactionError),
+    VerificationKey(ed25519_consensus::Error),
+    Verification(ed25519_consensus::Error),
+}
+
+/// A signed transaction.
+///
+/// [`SignedTransaction`] contains an [`UnsignedTransaction`] together
+/// with its signature and public key.
+#[derive(Clone, Debug)]
+pub struct SignedTransaction {
+    signature: Signature,
+    verification_key: VerificationKey,
+    transaction: UnsignedTransaction,
+}
+
+impl SignedTransaction {
+    #[must_use]
+    pub fn into_raw(self) -> raw::SignedTransaction {
+        let Self {
+            signature,
+            verification_key,
+            transaction,
+        } = self;
+        raw::SignedTransaction {
+            signature: signature.to_bytes().to_vec(),
+            public_key: verification_key.to_bytes().to_vec(),
+            transaction: Some(transaction.into_raw()),
+        }
+    }
+
+    /// Attempt to convert from a raw, unchecked protobuf [`raw::SignedTransaction`].
+    ///
+    /// # Errors
+    ///
+    /// Will return an error if signature or verification key cannot be reconstructed from the bytes
+    /// contained in the raw input, if the transaction field was empty (mmeaning it was mapped to
+    /// `None`), if the inner transaction could not be verified given the key and signature, or
+    /// if the native [`UnsignedTransaction`] could not be created from the inner raw
+    /// [`raw::UnsignedTransaction`].
+    pub fn try_from_raw(proto: raw::SignedTransaction) -> Result<Self, SignedTransactionError> {
+        use crate::Message as _;
+        let raw::SignedTransaction {
+            signature,
+            public_key,
+            transaction,
+        } = proto;
+        let signature =
+            Signature::try_from(&*signature).map_err(SignedTransactionError::signature)?;
+        let verification_key = VerificationKey::try_from(&*public_key)
```

Comment:
why the `&*public_key` instead of `public_key.as_slice()`?

Reply:
No reason other than it being more concise and . Unfortunately I cannot do `Signature::try_from(&signature)` because to make use of deref coercion because rust does not do deref coercion when matching traits (so the fact that `Vec<T>` derefs to`[T]` does not help). So I have to use`&*` to do that manually.